server/src/test/java/keywhiz/service/providers/ClientAuthFactoryTest.java

```diff
@@ -193,37 +206,71 @@ public void clientWhenPrincipalAbsentThrows() {
 
-    factory.provide(request, httpServletRequest);
+    when(xfccSourceConfig.callerSpiffeIdHeader()).thenReturn(callerSpiffeHeaderName);
+    when(request.getRequestHeader(callerSpiffeHeaderName)).thenReturn(List.of(clientSpiffe));
+
+    assertThat(factory.provide(request, httpServletRequest)).isEqualTo(client);
   }
 
-  @Test(expected = NotAuthorizedException.class)
-  public void rejectsXfcc_requesterNameNotAllowed() {
+  @Test
+  public void returnsClientWhenClientPresent_fromXfccHeader_customSpiffeIdHeader_invalidSpiffe() {
     when(httpServletRequest.getLocalPort()).thenReturn(xfccAllowedPort);
     when(request.getRequestHeader(ClientAuthFactory.XFCC_HEADER_NAME)).thenReturn(
         List.of(xfccHeader));
     when(securityContext.getUserPrincipal()).thenReturn(xfccPrincipal);
 
-    when(xfccSourceConfig.allowedClientNames()).thenReturn(List.of());
-    when(xfccSourceConfig.allowedSpiffeIds()).thenReturn(List.of(xfccSpiffe));
+    when(xfccSourceConfig.callerSpiffeIdHeader()).thenReturn(callerSpiffeHeaderName);
+    when(request.getRequestHeader(callerSpiffeHeaderName)).thenReturn(List.of("not-spiffe-id"));
+
+    assertThat(factory.provide(request, httpServletRequest)).isEqualTo(client);
+  }
+
+  @Test public void returnsNewClientWhenClientNotPresent_fromXfccHeader_customSpiffeIdHeader()
+      throws URISyntaxException {
+    when(httpServletRequest.getLocalPort()).thenReturn(xfccAllowedPort);
+    when(request.getRequestHeader(ClientAuthFactory.XFCC_HEADER_NAME)).thenReturn(
+        List.of(xfccHeader));
+    when(securityContext.getUserPrincipal()).thenReturn(xfccPrincipal);
+
+    when(xfccSourceConfig.callerSpiffeIdHeader()).thenReturn(callerSpiffeHeaderName);
+    when(request.getRequestHeader(callerSpiffeHeaderName)).thenReturn(List.of(newClientSpiffe));
+
+    // Create missing clients
+    when(clientAuthConfig.createMissingClients()).thenReturn(true);
+    // lookup doesn't find client
+    when(clientDAO.getClientByName(newClientName)).thenReturn(Optional.empty());
+    // a new DB record is created
+    when(clientDAO.createClient(eq(newClientName), eq("automatic"), any(),
+        eq(new URI(newClientSpiffe)))).thenReturn(123L);
+    when(clientDAO.getClientById(123L)).thenReturn(Optional.of(newClient));
+
+    assertThat(factory.provide(request, httpServletRequest)).isEqualTo(newClient);
+  }
+
+  @Test(expected = NotAuthorizedException.class)
+  public void rejectsXfcc_requesterAuthMissing() {
+    when(httpServletRequest.getLocalPort()).thenReturn(xfccAllowedPort);
+    when(request.getRequestHeader(ClientAuthFactory.XFCC_HEADER_NAME)).thenReturn(
+        List.of(xfccHeader));
+    when(securityContext.getUserPrincipal()).thenReturn(null);
 
     factory.provide(request, httpServletRequest);
   }
 
   @Test(expected = NotAuthorizedException.class)
-  public void rejectsXfcc_requesterSpiffeNotAllowed() {
+  public void rejectsXfcc_requesterNameNotAllowed_requesterSpiffeNotAllowed() {
```

Comment:
Does this test still need to be changed? The goal here iirc was to ensure that if the proxy provides both a name and SPIFFE ID, and _either_ of those is disallowed, then the request will be rejected even if the other was allowed. Unless we've changed that behavior, it should still be tested.

Reply:
My bad. I've changed it back